Prototype: Rely less on remote_task_check_interval.

anarazel opened the pull request:
When executing queries with `citus.task_executor = 'real-time'`, query execution could, so far, spend a significant amount of time sleeping. That's because we were

  * sleeping after several phases of query execution, even if we're not
waiting for network IO
  *  sleeping for a fixed amount of time when waiting for network IO;
often a lot longer than actually required

Just reducing the amount of time slept isn't a real solution, because that just increases CPU usage.

Instead have the real-time executor's `ManageTaskExecution` return whether a task is currently being processed, waiting for reads or writes, or failed. When all tasks are waiting for IO use `poll()` to wait for IO readiness.

That requires to slightly redefine how connection timeouts are handled: before we counted the number of times `ManageTaskExecution()` was called, and compared that with the timeout divided by the task check interval. That, if processing of tasks took a while, could significantly increase the time till a timeout occurred. Because it was based on the `ManageTaskExecution()` being called on a constant interval, this approach isn't feasible anymore. Instead measure the actual time since connection establishment was started. That could in theory, if task processing takes a very long time, lead to few passes over `PQconnectPoll()`.

The problem of sleeping too much also exists for the `task-tracker` executor, but is generally less problematic there, as processing the individual tasks usually will take longer. That said, for e.g. the regression tests it'd be helpful to use a similar approach.

# Review Tasks

  - [x] Rely on `ResetWaitInfo` to initialize new `WaitInfo`s
  - [x] Rename `RememberWait` to `MultiClientRegisterWait` (and update comment referencing the non-existent `AddWait` function)
  - [x] Remove extraneous `break`
  - [x] Consider `TaskConnStatus` instead of `WaitStatus`
  - [x] Consider `READY`, `ERRORED`, `WAIT_READ`, and `WAIT_WRITE` as enum states
  - [x] Add `INVALID` state, explicitly assign enum integer values
  - [x] Rename `WaitInfo` fields (`registeredWaits`, `maxWaits`, `connectionReady`, `errorOccurred`, `pollFileDescriptors`
  - [x] Rebase

ozgune commented:
Hey guys, I just went over this change and I'm adding a few notes.

1. @sumedhpathak will assign a reviewer to this pull request in the next two days. The only question in my mind was on ``ManageTaskExecution``'s API. I know @marcocitus thought about and changed this function's interface when he implemented the throttling changes. That said, this is so small that it shouldn't impact the reviewer.
2. Could we run example queries on 1-2 representative datasets and share our performance findings? What are the benefits say with 1-3 shards or 3-10 shards?
3. Since this is the default executor for Citus, could we test for node and task failures (task and fetch task)?
4. We conditionally ``#include <poll.h>``, ``<sys/poll.h>``, or ``<sys/select.h>`` in ``multi_client_executor.h``. Both the approach in this pull request and the one ``multi_client_executor.h`` work well. If these two files are consistent with each other, all the better.

marcocitus commented:
I think this approach is nice and simple. If network events come in one at a time: Are we going to run through all of the task executions for each of them. Is there an easy way to prevent that? (I'd be fine with following this approach even if there wasn't)

anarazel commented:
Yes, there's a way (not even that hard), but it'd make the patch more
complicated. I thought it'd be valuable to try first fix the biggest issue
at hand, without bigger refactoring. If we find that the repeated iteration
is a problem - something I've not managed in the scenarios I tested - we
can do some additional refactoring.

marcocitus commented:
Sounds good, I think we should get this into 5.2.

ozgune commented:
@jasonmp85 -- could you update the comments here as you review the changes?

If you run a few benchmarks, similar to the ones Onder ran for ``INSERT`` changes, could you also share them with the team?

anarazel commented:
> If you run a few benchmarks, similar to the ones Onder ran for ``INSERT`` changes, could you also share them with the team?

Isn't that a littlebit too much work for this?

ozgune commented:
Do you mean in proportion to the work we put in the change?

@samay-sharma earlier ran numbers for different ``remote_task_check_interval`` values on customer data, and it took half a day to gather the data. If we can quickly run these numbers - and it doesn't need to be @jasonmp85 who runs them - then we at least have input for our release notes?

marcocitus commented:
Wouldn't the benchmark just be: Run a SELECT * FROM empty_table, see the difference? Distributed query throughput is a bit of a hairy topic and unlikely to be affected by this change.

jasonmp85 commented on src/backend/distributed/executor/multi_client_executor.c:
```diff
@@ -653,6 +661,155 @@ MultiClientCopyData(int32 connectionId, int32 fileDescriptor)
+WaitInfo *
+MultiClientCreateWaitInfo(int maxConnections)
+{
+	WaitInfo *waitInfo = palloc(sizeof(WaitInfo));
```
Shouldn't you use `palloc0` here? The boolean fields aren't initialized to anything, seems better to just `palloc0` unless there's a clear reason not to.

anarazel replied:
Hm. I think a better solution would be to rely on ResetWaitInfo() here.

jasonmp85 replied:
Sounds good.

jasonmp85 commented on src/backend/distributed/executor/multi_client_executor.c:
```diff
@@ -653,6 +661,155 @@ MultiClientCopyData(int32 connectionId, int32 fileDescriptor)
+
+	waitInfo->allocated = maxConnections;
+	waitInfo->waiting = 0;
+	waitInfo->pollfds = palloc(maxConnections * sizeof(struct pollfd));
```
Same (`palloc0` unless proven otherwise).

anarazel replied:
That just hides bugs. As ->pollfd is reused, zeroing it in the first round, but not later, only serves to make issues harder to find.

jasonmp85 replied:
Wait, how does setting an array of pointers to zero hide bugs? Wouldn't the first attempt to use one segfault?

jasonmp85 commented on src/include/distributed/multi_client_executor.h:
```diff
@@ -72,6 +75,28 @@ typedef enum
 
 
+/* Enumeration to track what the task associated with a connection is waiting for */
+typedef enum WaitStatus
```
The names of these enums are really tripping me up. They're not always _track[ing] what the task associated with a connection is waiting for_, right? For instance, the error state is just a signal that an error has occurred, not e.g. that we're waiting for an error.

jasonmp85 replied:
Ok, how about calling this `TaskConnStatus`?

anarazel replied:
I don't like ConnStatus much, several things aren't related to the connection itself. What do you think about
```
/* Enumeration to track whether a task is ready to run and, if not, what it's blocked on*/
typedef enum TaskExecutionStatus
{
	TASK_STATUS_INVALID = 0,
	TASK_STATUS_ERROR, /* error occured */
	TASK_STATUS_READY, /* task ready to be processed further */
	TASK_STATUS_SOCKET_READ, /* waiting for connection to become ready for reads */
	TASK_STATUS_SOCKET_WRITE /* waiting for connection to become ready for writes */
} TaskExecutionStatus;
```

jasonmp85 replied:
Looks good. I hesitated to put _just_ the word read or write in the last two values, since I thought the word _need_ or_wait_ would provide more information. But this is better than `WaitStatus`! I think it's better enough.

anarazel replied:
On 2016-05-27 14:25:41 -0700, Jason Petersen wrote:
> > @@ -72,6 +75,28 @@ typedef enum
> >  } BatchQueryStatus;
> >  
> >  
> > +/* Enumeration to track what the task associated with a connection is waiting for */
> > +typedef enum WaitStatus
> 
> Looks good. I hesitated to put _just_ the word read or write in the last two values, since I thought the word _need_ or_wait_ would provide more information. But this is better than `WaitStatus`! I think it's better enough.

I think the SOCKET_ prefix makes that mostly clear.

jasonmp85 commented on src/backend/distributed/executor/multi_server_executor.c:
```diff
@@ -212,7 +212,7 @@ InitTaskExecution(Task *task, TaskExecStatus initialTaskExecStatus)
 	taskExecution->taskId = task->taskId;
 	taskExecution->nodeCount = nodeCount;
-	taskExecution->connectPollCount = 0;
+	taskExecution->connectStartTime = 0;
```
Since `TimestampTz` can be either an integer or double, can you add a cast to `TimestampTz` on the right hand side of this assignment?

anarazel replied:
It's a correct value for either, so ...?  We're not super strict in the codebase about adding casts for every promotion, and I don't really see the benefit to start with something as innocuous as above.

jasonmp85 commented on src/backend/distributed/executor/multi_real_time_executor.c:
```diff
@@ -357,6 +408,7 @@ ManageTaskExecution(Task *task, TaskExecution *taskExecution)
 			else
 			{
 				taskStatusArray[currentIndex] = EXEC_TASK_FAILED;
+				break;
```
Think this is extraneous.

jasonmp85 commented on src/include/distributed/multi_client_executor.h:
```diff
@@ -72,6 +75,28 @@ typedef enum
+
+typedef struct WaitInfo
+{
+	int allocated;
```
Isn't this the length of the `pollfds` array? For convention's sake can we move it to be adjacent to that and rename it to be more clear (`pollfdcount`?). Even though the array isn't inline in the `struct`, I'm still used to seeing lengths next to the end-of-struct arrays.

jasonmp85 commented on src/include/distributed/multi_client_executor.h:
```diff
@@ -72,6 +75,28 @@ typedef enum
+typedef struct WaitInfo
+{
+	int allocated;
+	int waiting;
```
And isn't this more like "registered waits"? `waiting` is not very descriptive here.

jasonmp85 commented on src/backend/distributed/executor/multi_client_executor.c:
```diff
@@ -653,6 +661,155 @@ MultiClientCopyData(int32 connectionId, int32 fileDescriptor)
+ * waitStatus.
+ */
+void
+MultiClientRememberWait(WaitInfo *waitInfo, WaitStatus waitStatus, int32 connectionId)
```
What do you think about `MultiClientRegisterWait`? I noticed in another comment you referenced `MultiClientAddWait`. Is there a reason that name was abandoned? `Remember` is a weird way of saying "add to array".

anarazel replied:
RegisterWait works for me.

jasonmp85 commented on src/include/distributed/multi_client_executor.h:
```diff
@@ -72,6 +75,28 @@ typedef enum
+/* Enumeration to track what the task associated with a connection is waiting for */
+typedef enum WaitStatus
+{
+	WAIT_BUSY, /* no IO required, ready to be processed further */
```
And here we'd have `TASK_CONN_READY`, `TASK_CONN_ERRORED`, `TASK_CONN_WAIT_READ`, and `TASK_CONN_WAIT_WRITE`

jasonmp85 replied:
I'd be remiss not to point out that (a) house style is to reserve the first enumeration for an explicit "invalid" value, and (b) house style is to explicitly assign each enum. There are arguments on either side of this, but I'd rather have a consistent codebase we can swap to another style than have a codebase whose style varies.

anarazel replied:
On 2016-05-27 10:00:26 -0700, Jason Petersen wrote:
> I'd be remiss not to point out that (a) house style is to reserve the
> first enumeration for an explicit "invalid" value, and (b) house style
> is to explicitly assign each enum. There are arguments on either side
> of this, but I'd rather have a consistent codebase we can swap to
> another style than have a codebase whose style varies.

Gah, these are *bad* policies, especially the second one.

jasonmp85 replied:
Not a fan either, but like I said, let's change them in one fell swoop and not by introducing inconsistencies until the original style "ages out".

anarazel replied:
> Not a fan either, but like I said, let's change them in one fell swoop and not by introducing inconsistencies until the original style "ages out".

Meh, that just leads to never changing anything, because a
one-big-change is too much work/churn.  There's a lot of not so optimal
policies in the code-base and afaics not one has been fixed as a whole.
With a codebase increasing in size that's going to be less and less realistic.

jasonmp85 replied:
I think the problem here is that the invalid value requirement comes out of the "initialize variables" requirement.

I'm not really sure where the "assign an explicit value" requirement comes from (other than the fact that, I believe, C doesn't actually say what the values will be in the spec?)

jasonmp85 replied:
Anyways, I guess leave it as-is. Incremental change could work, too, I suppose.

jasonmp85 commented on src/include/distributed/multi_client_executor.h:
```diff
@@ -72,6 +75,28 @@ typedef enum
+	int waiting;
+	bool haveBusy;
+	bool haveFailure;
+	struct pollfd *pollfds;
```
And I think I'd understand:

```c
int registeredWaits;
int maxWaits;
bool haveReady;
bool errorOccurred;
struct pollfd *pollFileDescriptors;
```

jasonmp85 commented:
This looks good to me. The only thing that really threw me when reading it was the `WaitStatus` value names. Only two of them trigger a wait, the others signal we're ready to immediately move ahead because of readiness or error. I tried to come up with a better name, but it's pretty hard… the best I could devise was `TaskConnStatus`, and I think `WAIT_BUSY` and `WAIT_ERROR` should be `TASK_CONN_READY` and `TASK_CONN_ERRORED`.

Additionally, the fields in `WaitInfo` could use some clarification; I posted a suggestion.

Beyond that, the structure of this looks great. Can you address the feedback, rebase, and I'll run through some testing before a merge?

jasonmp85 commented:
Context dump: We discussed this PR at length in chat yesterday and the regression test failures would be better handled in whatever solution comes out of #345. This PR is complete as-is and represents an improvement for latency of these tasks.

anarazel commented:
Pushed a rebased version, after addressing most of the comments. I left the variable name at pollfd, and chose a slightly different name (discussed above), instead of TaskConnStatus.

jasonmp85 commented:
My testing didn't show any problems with this; still distilling a simple benefit scenario, but I think this is good to :shipit:!

anarazel commented:
On 2016-06-02 11:04:03 -0700, Jason Petersen wrote:
> My testing didn't show any problems with this; still distilling a simple benefit scenario, but I think this is good to :shipit:!

Any relatively simple multi-shard query should do. Or even a single shard
one, if it's not router capable.

Thanks for the review.

jasonmp85 commented:
> Any relatively simple multi-shard query should do. Or even a single shard one, if it's not router capable.

Sorry, I mean something with a pathologically large divergence (like a query that runs _just_ over the old timeout time), where old way would be around two times slower than the new way.

But yeah, I see what you mean.
